README.md
# X-Plane Aircraft Information Plugin

## About

The Aircraft Information (ACInfo) plugin aims to make it easy to 
identify the name and type of an X-Plane aircraft for other applications
and plugins by providing custom datarefs that encode this information.

### DataRefs

ACInfo provides the following DataRefs:

*   `vscl/sim/aircraft/name`: the aircraft name as a string (`char *`).
*   `vscl/sim/aircraft/type`: the aircraft type as an `int64_t` bitmask.

These can be requested using the X-Plane SDK or the new DREF UDP 
interface.

### Types

Types are a generic representation of what an aircraft is.
They include powerplant type and aircraft role.
Since these can be combined, types are represented as a bitmask, so e.g.
a jet fighter would be represented as `JET ^ FIGHTER`.
Refer to the [`lua/acinfo.lua`](lua/acinfo.lua) script for the most
up-to-date type definitions.

### Algorithms

(Well, they're sort of algorithms...)
ACInfo uses the aircraft's ICAO name (`sim/aircraft/view/acf_ICAO`) if
available as the aircraft name.
If this is known to be empty or bad, the aircraft's filename is used.
In some cases, this is then mapped to a custom name.

To determine engine type, the plugin looks at engine DataRefs specific 
to piston-engined aircraft and sees if they are nonzero.
If prop RPM and manifold pressure are both nonzero, the aircraft is 
determined to be piston-engined, and if prop RPM only is nonzero, the
aircraft is a turboprop.
Otherwise, it is considered a jet (inclusive of turbojets and 
turbofans).
For the aircraft role, the aircraft's path is iterated over to find the
common X-Plane subdirectories, which are then mapped to a type integer.
In certain cases this is overriden with known types.
The Lua script assumes the path convention used in the VSCL for 
additional aircraft (e.g. from X-Plane.org); you will likely need to 
customize this for your X-Plane deployment.

## Requirements

ACInfo implements most of its core logic in the Lua programming 
language.
Starting with X-Plane 10.20, Lua is embedded into X-Plane and can be 
used to script the simulator.
ACInfo uses the [FlyWithLua](https://github.com/X-Friese/FlyWithLua)
plugin to provide a convenient API to interact with X-Plane.
Accordingly, FlyWithLua must be installed for X-Plane.

For the C plugin, the 
[X-Plane SDK](http://www.xsquawkbox.net/xpsdk/mediawiki/Main_Page)
is required.
The SDK is included as a submodule with the path `ext/sdk`, and can
be instantiated with
```
git submodule update --init
```
which will initialize and update all of the submodules.

Additionally, `libbsd` is required as the plugin makes use of 
`strlcpy(3)`.

## Building the C Plugin

ACInfo uses [CMake](https://cmake.org) as its build system and is 
designed for an out-of-source build.

To build the plugin, enter the following commands in the project
root directory:
```
mkdir build && pushd build
cmake ..
make
popd
```
which will output the plugin to `lib/libacinfo.so`.
This file will be symlinked to [`64/lin.xpl`](64/lin.xpl) for use as an
X-Plane "fat plugin".

## Installing the Plugin 
### C Plugin 
Place the plugin into the `Resources/plugins` directory of your X-Plane
installation with the following structure:
```
acinfo/
├── 64
│   └── lin.xpl
└── lib
    └── libacinfo.so
```

### Lua Scripts
To install the Lua script(s), place them into the `Scripts` directory
of the main FlyWithLua plugin directory.

## License
See the [`LICENSE.md`](LICENSE.md) file for license and copyright 
information.


changelog.txt
* Thu Jun 23 2016 Joshua Harris <joshua.a.harris@tamu.edu> 1.0.0-1
- Initial RPM release


src/acinfo.c
/*
 * acinfo: X-Plane aircraft info plugin.
 * See LICENSE file for copyright and license details.
 *
 * acinfo.c: Main X-Plane plugin implementation.
 */
#include "acinfo.h"

int
XPluginStart(char *out_name, char *out_sig, char *out_desc)
{
	char buf[DEFAULT_BUFSIZE];

	/* Add plugin information. */
	strlcpy(out_name, name, strlen(name) + 1);
	strlcpy(out_sig, sig, strlen(sig) + 1);
	strlcpy(out_desc, desc, strlen(desc) + 1);

	/*
	 * Print loading information.
	 * Return value is casted to void since we don't really care about
	 * truncation; indeed the default bufsize should always be larger than
	 * the input string.
	 */
	(void) snprintf(buf, DEFAULT_BUFSIZE, "%s: version %s loaded.\n",
	                short_name, _VERSION);
	XPLMDebugString(buf);

	/* Register datarefs. */
	/*@-nullpass@*/
	aircraft_name = XPLMRegisterDataAccessor(aircraft_name_dr,
	                                         xplmType_Data,
	                                         XPDREF_WRITABLE,
	                                         NULL, NULL,
	                                         NULL, NULL,
	                                         NULL, NULL,
	                                         NULL, NULL,
	                                         NULL, NULL,
	                                         get_acname, set_acname,
	                                         NULL, NULL);
	aircraft_type = XPLMRegisterDataAccessor(aircraft_type_dr,
	                                         xplmType_Int,
	                                         XPDREF_WRITABLE,
	                                         get_actype, set_actype,
	                                         NULL, NULL,
	                                         NULL, NULL,
	                                         NULL, NULL,
	                                         NULL, NULL,
	                                         NULL, NULL,
	                                         NULL, NULL);
	/*@=nullpass@*/

	/* Initialize datarefs. */
	memset(aircraft_name_value, 0, sizeof(char) * ACNAME_BUFSIZE);
	strlcpy(aircraft_name_value, default_name, strlen(default_name) + 1);
	aircraft_type = XPLMFindDataRef(aircraft_type_dr);
	aircraft_name = XPLMFindDataRef(aircraft_name_dr);

	/* Register the datarefs with the DataRef Editor. */
	/*@-nullpass@*/
	XPLMRegisterFlightLoopCallback(register_dr, 1, NULL);
	/*@=nullpass@*/

	return PLUGIN_LOAD_SUCCESS;
}

void
XPluginStop(void)
{
	/* Unregister everything. */
	/*@-nullpass@*/
	XPLMUnregisterFlightLoopCallback(register_dr, NULL);
	XPLMUnregisterDataAccessor(aircraft_name);
	XPLMUnregisterDataAccessor(aircraft_type);
	/*@=nullpass@*/
}

int
XPluginEnable(void)
{
	return PLUGIN_ENABLE_SUCCESS;
}

void
XPluginDisable(void)
{
	return;
}

void
XPluginReceiveMessage(/*@unused@*/ UNUSED XPLMPluginID src,
                      /*@unused@*/ UNUSED int msg,
                      /*@unused@*/ UNUSED void* params)
{
	return;
}

static int
get_acname(/*@unused@*/ UNUSED void *refcon, void *out, int offset, int max)
{
	char tmp[ACNAME_BUFSIZE];
	size_t maxsz, n, name_len;

	/* If no output pointer is given, return the array buffer size. */
	if (out == NULL) {
		return ACNAME_BUFSIZE;
	}

	/* Make the max size a size_t. */
	maxsz = (size_t) max;

	/* Calculate number of bytes to return. */
	name_len = strlen(aircraft_name_value);
	n = name_len + 1 - offset;
	if (n > maxsz) {
		n = maxsz;
	}

	/* Copy bytes and null-terminate. */
	strlcpy(tmp, aircraft_name_value, n);
	memcpy(out, (void *) &tmp, n);
	return (int) n;
}

static int
get_actype(/*@unused@*/ UNUSED void *refcon)
{
	return aircraft_type_value;
}

static float
register_dr(/*@unused@*/ UNUSED float call_time,
            /*@unused@*/ UNUSED float elapsed_time,
            /*@unused@*/ UNUSED int ctr,
            /*@unused@*/ UNUSED void *refcon)
{
	XPLMPluginID pid = XPLMFindPluginBySignature(dresig);
	if (pid != XPLM_NO_PLUGIN_ID) {
		char buf[DEFAULT_BUFSIZE];
		XPLMSendMessageToPlugin(pid, MSG_ADD_DATAREF,
		                        (void *) aircraft_type_dr);
		(void) snprintf(buf, DEFAULT_BUFSIZE,
			        "%s: dataref \"%s\" registered.\n",
		                short_name, aircraft_type_dr);
		XPLMDebugString(buf);
		XPLMSendMessageToPlugin(pid, MSG_ADD_DATAREF,
		                        (void *) aircraft_name_dr);
		(void) snprintf(buf, DEFAULT_BUFSIZE,
		                "%s: dataref \"%s\" registered.\n",
		                short_name, aircraft_name_dr);
		XPLMDebugString(buf);
	}
	return 0;
}

static void
set_acname(/*@unused@*/ UNUSED void *refcon, void *in, int offset, int max)
{
	size_t maxsz, n;

	/* Make the max size a size_t. */
	maxsz = (size_t) max;

	/* Calculate number of bytes to copy. */
	n = ACNAME_BUFSIZE - (size_t) offset;
	if (n > maxsz) {
		n = maxsz;
	}

	/* Copy bytes and ensure null-termination. */
	memcpy(aircraft_name_value, (char *) in, n - 1);
	aircraft_name_value[n] = '\0';
}

static void
set_actype(/*@unused@*/ UNUSED void *refcon, int in)
{
	aircraft_type_value = in;
}


include/acinfo.h
/*
 * acinfo: X-Plane aircraft info plugin.
 * See LICENSE file for copyright and license details.
 *
 * acinfo.h: Main X-Plane plugin.
 */
#pragma once

#include <bsd/string.h>
#include <stdio.h>
#include <string.h>

#include "XPLMDataAccess.h"
#include "XPLMPlugin.h"
#include "XPLMProcessing.h"
#include "XPLMUtilities.h"

/*
 * Define the unused attribute to suppress warnings about unused parameters in
 * functions the X-Plane SDK requires but this plugin does not use.
 */
#if (defined __GNUC__ || defined __GNUG__ || defined __clang__)
#	define UNUSED __attribute__((__unused__))
#else
#	define UNUSED
#endif

/* Array sizes. */
#define DEFAULT_BUFSIZE 512
#define ACNAME_BUFSIZE 1024

/* Constants for DRE. */
#define XPDREF_WRITABLE 1
#define MSG_ADD_DATAREF 0x01000000

/* Define constants to avoid X-Plane SDK magic numbers. */
#define PLUGIN_LOAD_SUCCESS 1
#define PLUGIN_ENABLE_SUCCESS 1

/* Plugin description. */
static const char *name = "Aircraft Information Plugin";
static const char *short_name = "acinfo";
static const char *sig = "edu.tamu.vscl.acinfo";
static const char *desc = "Provides custom datarefs describing aircraft.";

/* Other plugin signatures. */
static const char *dresig = "xplanesdk.examples.DataRefEditor";

/* Custom datarefs. */
static const char *aircraft_type_dr = "vscl/sim/aircraft/type";
static const char *aircraft_name_dr = "vscl/sim/aircraft/name";
/*@null@*/ static XPLMDataRef aircraft_type = NULL;
/*@null@*/ static XPLMDataRef aircraft_name = NULL;

/* Values to hold the data. */
static int aircraft_type_value = 0;
/*@unique@*/ static char aircraft_name_value[ACNAME_BUFSIZE];
static const char *default_name = "unknown";

/* Getters/setters for datarefs. */
static int get_acname(UNUSED void *, void *, int, int);
static int get_actype(UNUSED void *);
static void set_acname(UNUSED void *, void *, int, int);
static void set_actype(UNUSED void *, int);

/* Deferred initialization loop for registering with DataRef Editor. */
static float register_dr(UNUSED float, UNUSED float, UNUSED int, UNUSED void *);


lua/acinfo.lua
--[[
    acinfo: X-Plane aircraft info plugin.
    See LICENSE file for copyright and license details.

    acinfo.lua: Sets aircraft type and name in FlyWithLua.
--]]
require("bit")
require("string")


-- Acquire the VSCL custom datarefs.
DataRef("ac_name", "vscl/sim/aircraft/name", "writable", 0)
DataRef("ac_type", "vscl/sim/aircraft/type", "writable")
DataRef("prop_rpm", "sim/cockpit2/engine/indicators/prop_speed_rpm",
        "readonly", 0)
DataRef("manifold_press", "sim/cockpit2/engine/indicators/MPR_in_hg",
        "readonly", 0)


-- Define an array of aircraft types. These values are used as a bitmask.
local aircraft_types = {
    ["unknown"] = 0,
    ["jet"] = 1,
    ["piston"] = 2,
    ["turboprop"] = 4,
    ["airliner"] = 8,
    ["attack"] = 16,
    ["cargo"] = 32,
    ["bomber"] = 64,
    ["ew"] =128,
    ["fighter"] = 256,
    ["ga"] = 512,
    ["glider"] = 1024,
    ["helicopter"] = 2048,
    ["recon"] = 4096,
    ["seaplane"] = 8192,
    ["spacecraft"] = 16384,
    ["tanker"] = 32768,
    ["uav"] = 65536,
    ["vtol"] = 131072,
    ["xplane"] = 262144
}


-- Define an array of path names mapping to the above types.
local type_from_path = {
    ["Attackers"] = "attack",
    ["Bombers"] = "bomber",
    ["EW"] = "ew",
    ["Fighters"] = "fighter",
    ["General Aviation"] = "ga",
    ["Gliders"] = "glider",
    ["Heavy Metal"] = "airliner",  -- Assume for now that HM is an airliner.
    ["Helicopters"] = "helicopter",
    ["Mega-Planes!"] = "bomber", -- Assume for now that MP! are bombers.
    ["Radio Control"] = "unknown", -- We don't care about RC aircraft.
    ["Science Fiction"] = "fighter", -- AFAIK, this is just the anime fighter.
    ["Seaplanes"] = "ga", -- It's sort of GA.
    ["Space Ships"] = "spacecraft",
    ["Tankers"] = "tanker",
    ["Transports"] = "cargo",
    ["UAS"] = "uav",
    ["VTOL"] = "vtol",
    ["X-Planes"] = "xplane"
}


-- Checks to see if a string is empty.
function isempty(s)
    return s == nil or s == ''
end


-- from http://lua-users.org/wiki/SplitJoin
function string:split(sSeparator, nMax, bRegexp)
    assert(sSeparator ~= '')
    assert(nMax == nil or nMax >= 1)

    local aRecord = {}

    if self:len() > 0 then
        local bPlain = not bRegexp
        nMax = nMax or -1

        local nField, nStart = 1, 1
        local nFirst,nLast = self:find(sSeparator, nStart, bPlain)
        while nFirst and nMax ~= 0 do
            aRecord[nField] = self:sub(nStart, nFirst-1)
            nField = nField+1
            nStart = nLast+1
            nFirst,nLast = self:find(sSeparator, nStart, bPlain)
            nMax = nMax-1
        end
        aRecord[nField] = self:sub(nStart)
    end

    return aRecord
end


-- Determine the best name for the aircraft.
function get_acname()
    if isempty(PLANE_ICAO) then
        local name = string.gsub(AIRCRAFT_FILENAME, ".acf", "")
        -- TODO: Add mapping from filename to a "human-friendly" name.
        ac_name = name
    else
        ac_name = PLANE_ICAO
    end
end


-- Determine the type of the aircraft as best as possible.
function get_actype()
    ac_type = aircraft_types["unknown"]
    --[[
        Determine if the aircraft is jet powered by seing if the manifold
        pressure and propeller RPM are nonzero. A piston-engined aircraft has
        both nonzero and a turboprop has RPM nonzero.
    --]]
    if prop_rpm > 0 then
        if manifold_press > 0 then
            ac_type = bit.bxor(ac_type, aircraft_types["piston"])
        else
            ac_type = bit.bxor(ac_type, aircraft_types["turboprop"])
        end
    else
        ac_type = bit.bxor(ac_type, aircraft_types["jet"])
    end
    --[[
        Use the file path of the aircraft file to determine the type based on
        common X-Plane numbering schemes.
    --]]
    for _, v in pairs(string.split(AIRCRAFT_PATH, "/")) do
        local tmp = type_from_path[v]
        if tmp ~= nil then
            ac_type = bit.bxor(ac_type, aircraft_types[tmp])
        end
    end
end


-- Update the name and type of aircraft. We should only need to do this on
-- aircraft loads.
function main()
    -- Write the current values to the log file.
    logMsg(string.format("ACInfoLua Info: current name <%s>", ac_name))
    logMsg(string.format("ACInfoLua Info: current type <%s>", ac_type))

    -- Update the datarefs.
    get_acname()
    get_actype()

    -- Log the updated values.
    logMsg(string.format("ACInfoLua Info: name set to <%s>", ac_name))
    logMsg(string.format("ACInfoLua Info: type set to <%s>", ac_type))
end


-- Run the main function.
main()
